📖 CAEP: Flexible Managed Kubernetes Endpoints (#8500)
kubernetes-sigs/cluster-api: docs/proposals/20230407-flexible-managed-k8s-endpoints.md

kind: ClusterEndpoint
metadata:
   labels:
      cluster.x-k8s.io/cluster-name: my-cluster
spec:
  cluster: my-cluster
  host: "my-cluster-1234567890.region.elb.amazonaws.com"
  port: 1234
  type: ExternalControlPlaneEndpoint
```

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterEndpoint
metadata:
   labels:
      cluster.x-k8s.io/cluster-name: my-cluster-2
spec:
  cluster: my-cluster-2
  host: "10.40.85.102"
  port: 1234
  type: ExternalControlPlaneEndpoint
```

This is how the type specification would look:

```go
// ClusterEndpointType describes the type of cluster endpoint.
type ClusterEndpointType string

// ClusterEndpoint represents a reachable Kubernetes API endpoint serving a particular cluster function.
type ClusterEndpoint struct {
  metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

  Spec ClusterEndpointSpec   `json:"spec,omitempty"`
}

// ClusterEndpointSpec defines the desired state of the Cluster endpoint.
type ClusterEndpointSpec struct {
	// The Host is the DNS record or the IP address that the endpoint is reachable on.
	Host string `json:"host"`

	// The port on which the endpoint is serving.
	Port int32 `json:"port"`

>   // Cluster is a reference to the cluster name that this endpoint is reachable on.

Reviewer: nit: ident is off

One particular part of the existing Cluster API surface area that is inconsistent with most Managed Kubernetes user experiences is the accounting of the [Kubernetes API server](https://kubernetes.io/docs/concepts/overview/components/#kube-apiserver). In the canonical "self-managed" user story that Cluster API addresses, it is the provider implementation of Cluster API (e.g., CAPA) that is responsible for scaffolding the necessary _Kubernetes Cluster Infrastructure_ that is required in order to create the Kubernetes API server (e.g., a Load Balancer and a public IP address). This provider responsibility is declared in the `<Infra>Cluster` resource, and carried out via its controllers; and then finally this reconciliation is synchronized with the parent `Cluster` Cluster API resource.

> Because there exist Managed Kubernetes scenarios that handle a subset or all _Kubernetes Cluster Infrastructure_ responsibilities themselves, Cluster API's requirement of a `<Infra>Cluster` resource leads to undesirable implementation decisions, because in these scenarios there is no actual work for a Cluster API provider to do to scaffold _Kubernetes Cluster Infrastructure_.

Reviewer 2: This is only true in the case where no additional infra is required (i.e. all required infra is created when creating an instance of the service providers managed k8s offering). For example, GKE and EKS do not fall into this scenario.

Reviewer 2: That clarifies it well, thanks @jackfrancis

### Goals

- Build upon [the existing Cluster API Managed Kubernetes proposal](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220725-managed-kubernetes.md). Any net new recommendations and/or proposals will be a continuation of the existing proposal, and consistent with its original conclusions.

> - Identify and document API changes and controllers changes required to omit the `<Infra>Cluster` entirely.

Reviewer 2: ```suggestion
- Identify and document API changes and controllers changes required to omit the `<Infra>Cluster` entirely, where this is applicable.
```

### Design

Below we are documenting API changes and controllers changes required to omit the `<Infra>Cluster` entirely and to allow the `ControlPlane Provider` component  to take ownership of the responsibility of creating the control plane endpoint.

#### Core Cluster API changes

This proposal does not introduce any breaking changes for the existing "core" API. More specifically:

The existing Cluster API types are already able to omit the `<Infra>Cluster`:

- The `infrastructureRef` field on the Cluster object is already a pointer and thus it could be set to nil, and in fact we are already creating Clusters without `infrastructureRef` when we use a cluster class).
- The `infrastructure.Ref` field on the ClusterClass objects already a pointer and thus it could be set to nil, but in this case it is required to change the validation webhook to allow the user to not specify it; on top of that, when validating inline patches, we should reject patches targeting the infrastructure template objects if not specified.

> In order to allow the `ControlPlane Provider` component  to take ownership of the responsibility of creating the control plane endpoint we are going to introduce a new `ClusterEndpoint` CRD, below some example:

Reviewer 3: Given the questions below, could we have a section with a Go struct that describes the CRD?

Author: I've added that, PTAL

Reviewer 3: I might have missed it, I don't see it the `ClusterEndpoint` CRD struct?

Author: L187 at present

```yaml
apiVersion: cluster.x-k8s.io/v1beta1

> kind: ClusterEndpoint

Reviewer 3: I suggested ClusterEndpoint, given that we can eventually expand it with a `type` field of sorts to determine what it applies to

> kind: ClusterEndpoint

Reviewer 4: +1 to keep it generic, so in the future this could cover also a use case different from the current ControlPlaneEndpoint (e.g internal + external control plane endpoints).

> kind: ClusterEndpoint

Author: thanks everyone for weighing, is that sufficient to address your concerns @CecileRobertMichon?

> kind: ClusterEndpoint

Author: Super, thanks!

spec:
  hostname: 'name-1234567890.region.elb.amazonaws.com'

>   port: 1234

Reviewer: it might be helpful to define the actual types in the proposal instead of just examples

Author: done

  type: ExternalControlPlaneEndpoint
```

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterEndpoint
spec:

>   ip: 10.40.85.102

Reviewer 5: For this API to be future proof are we planning to make ip and port slices?
Or should this be an slice of endpoints so in future each element of the slice can have a type?

Reviewer: That's an interesting question... Maybe each endpoint should be its own object and have an owner ref to the control plane object?

Reviewer 3: This should probably be ~close to what's in https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.Listener 

A slice of ~Listeners (like in gateway networking) with a hostname and port should be sufficient to describe the endpoints?

Author: I think we want to only have one IP and port per ClusterEndpoint (singular endpoint).

Should we update the spec to include an ownerRef?

```
spec:
...
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: AWSManagedControlPlane
    name: my-eks-cluster
    namespace: my-eks-cluster-ns
```

or

```
spec:
...
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
    kind: AzureManagedControlPlane
    name: my-aks-cluster
    namespace: my-aks-cluster-ns
```

Reviewer 6: Not sure if it's a good idea to add a ref in spec as ownerRef if Kubernetes has metadata.ownerRef for this purpose

Reviewer 3: We can have it as a slice and limit to a single one via a webhook or openapi validation, that opens the possibility later to handle more than one

Author: What are we planning to make slice of? We can't make both host and port slices, as there are 1:1 relationships between the two of those properties. IMO any "set of endpoints" implementations can employ slices of this new type, rather than having a `ClusterEndpoint` (singular) represents sets of hosts and ports.

WDYT?

Reviewer 3: A slice of `listeners`? https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.Listener

Author: Thx. Can someone explain how "a singular endpoint containing multiple endpoints" is preferable to "multiple endpoints"? Why would an endpoint have multiple endpoints? (semantically speaking, a "listener" in this case is just a synonym for "endpoint")

Reviewer 4: my two cents is to keep it simple because this covers the use case in the scope of the current work (the control plane endpoint) and unblock this work focused on managed K8s enablement.
Considering that this will land in 1.7, there are 4-5 months of time to do another iteration on this proposal/add another proposal extending this CR making it possible to support more control endpoints/more complex types of endpoints

  port: 1234
  type: ExternalControlPlaneEndpoint
```

This type will be the primary way for both `<Infra>Cluster` or `ControlPlane Provider` to communicate the ClusterAddress to "core" Cluster API controllers.

The `<Infra>Cluster` object which is currently using the `spec.controlPlaneEndpoint` for the same scope will continue to work because "core" Cluster API controllers will continue to recognize when this field is set and take care of generating the `ClusterEndpoint` automatically; however this mechanism should be considered as a temporary machinery to migrate to the new CRD, and it will be removed in future versions of Cluster API. In addition, once the legacy behavior is removed, we will deprecate and eventually remove the `spec.controlPlaneEndpoint` field from the `Cluster` CustomResourceDefinition, and recommend that providers do the same for their `<Infra>Cluster` CustomResourceDefinitions as well.

Notes:

> - The `hostname` and `ip` fields are mutually exclusive

Reviewer: the current struct defined in CAPI for API Endpoint just calls this host and it can be either the DNS name or IP address, might make conversion easier to keep it the same

Reviewer 3: +1, a host can be both a dns name or ip

Author: changed to host

> - The `type` field has been introduced to allow to extend the usage of this CRD to address https://github.com/kubernetes-sigs/cluster-api/issues/5295 in a future iterations

Reviewer 5: I'm not sure I'm seeing how introducing a "type" now enables the use case any better than not introducing a type at all for now. I think we'll be in a better position to reason about the best way to add it when we actually need to disambiguate between types.
By introducing it now I think we'll add friction for existing use cases which might implement a private topology, in which case forcing them to have an endpoint of type ExternalControlPlaneEndpoint seems inappropriate.

Reviewer 3: We can add it later probably, and default it. Adding it now might make it explicit, and less to deal with conversion later

Author: If we don't add it now (and a type=ControlPlaneEndpoint is implied), then adding it later should allow us to easily make that default assignment (if type=="" then set type=ControlPlaneEndpoint) in future versions for back-compat.

Do we have consensus that such an approach (wait to introduce until we have a > 1 type use case) is preferable to explicitly including a new `type` property in the scope of this PR?

Reviewer 6: I'm fine with adding it later until we have a better idea about other values

Reviewer 3: Later sounds good

Author: Done

#### Infra Providers API changes

This proposal does not introduce any breaking changes for the provider's API.

However, Infra providers MUST deprecate the `spec.controlPlaneEndpoint` field in their `<Infra>Cluster` so they can remove it in a future API version.

#### Core Cluster API Controllers changes

- All the controllers working with ClusterClass objects must take into account that the `infrastructure.Ref` field could be omitted; most notably:
  - The ClusterClass controller must ignore nil `infrastructure.Ref` fields while adding owner references to all the objects referenced by a ClusterClass.
  - The Topology controller must skip the generation of the `<Infra>Cluster` objects when the `infrastructure.Ref` field in a ClusterClass is empty.

- All the controllers working with Cluster objects must take into account that the `infrastructureRef` field could be omitted; most notably:
  - The Cluster controller must use skip reconciling this external reference when the `infrastructureRef` is missing; also, the `status.InfrastructureReady` field must be automatically set to true in this case.

> - The Cluster controller must reconcile the new `ClusterEndpoint` CR. Please note that:

Reviewer: This is an implementation detail but since it's specified here, why not have a separate controller for the new CRD?

Reviewer 4: This is definitely an option, but let's defer this to the implementation.
I just checked and the current implementation of the cluster controller already handles a couple of nested reconcile functions, and thus this one can also (alternative to seprated controller) fit into the same pattern

https://github.com/fabriziopandini/cluster-api/blob/f0b2b03bc7cd35d259a64954bb03c66dc4155700/internal/controllers/cluster/cluster_controller.go#L194-L197

Reviewer: + to deferring, in this case let's not specify which controller in the proposal?

Author: That is sensible. Updated, PTAL.

  - The value from the `ClusterEndpoint` CRD must surface on the `spec.ControlPlaneEndpoint` field on the `Cluster` object.
  - If both are present, the value from the `ClusterEndpoint` CRD must take precedence on the value from `<Infra>Cluster` objects still using the `spec.controlPlaneEndpoint`.

- The Cluster controller must implement the temporary machinery to migrate to the new CRD existing Clusters and to deal with `<Infra>Cluster` objects still using the `spec.controlPlaneEndpoint` field as a way to communicate the ClusterAddress to "core" Cluster API controllers:
  - If there is the `spec.ControlPlaneEndpoint` on the `Cluster` object but not a corresponding `ClusterEndpoint` CR, the CR must be created.

#### Provider controller changes

- All the `<Infra>Cluster` controller who are responsible to create a control plane endpoint
  - As soon as the `spec.controlPlaneEndpoint` field in the `<Infra>Cluster` object will removed, the `<Infra>Cluster` controller must instead create a `ClusterEndpoint` CR to communicate the control plane endpoint to the Cluster API core controllers

>     - NOTE: technically it is possible to start creating the `ClusterEndpoint` CR *before* the removal of the `spec.controlPlaneEndpoint` field, because the new CR will take precedence on the value read from the field, but this is up to the infra provider maintainers.

Reviewer: I think this contradicts the earlier statement

> Infra providers MUST deprecate the `spec.controlPlaneEndpoint` field in their `<Infra>Cluster` so they can remove it in a future API version

Reviewer 3: Not sure if this is a `MUST` really, we can suggest, and provide a warning to users/providers

Author: Updated the language in the earlier statement

  - The `ClusterEndpoint` CR must have an owner reference to the `<Infra>Cluster` object from which it is originated.

> - All the `ControlPlane Provider` controller who are responsible to create a control plane endpoint

Reviewer: ```suggestion
- All the `ControlPlane Provider` controllers who are responsible for creating a control plane endpoint
```

Author: thx!

>   - Must stop to wait for the `spec.ControlPlaneEndpoint` field on the `Cluster` object to be set before starting to provision the control plane.

Reviewer 2: I may be reading this wrong but i'm failing to understand this. 

Is this saying that the `AzureManagedControlPlane/AWSManagedControlPlane` reconciliation must wait until the `spec.ControlPlaneEndpoint` on **Cluster** is set?

Reviewer 2: I would've thought we need flip this, so it becomes:

- Must not wait for the `spec.ControlPlaneEndpoint` field on the `Cluster` object to be set before starting to provision the control plane.

Reviewer: ```suggestion
  - Must no longer to wait for the `spec.ControlPlaneEndpoint` field on the `Cluster` object to be set before starting to provision the control plane.
```

might be more clear

Author: I would actually like @fabriziopandini to weigh in here :)

Reviewer 4: I like Cecile phrasing, +1

>   - Must stop to wait for the `spec.ControlPlaneEndpoint` field on the `Cluster` object to be set before starting to provision the control plane.

Reviewer: ```suggestion
  - Must wait for the `spec.ControlPlaneEndpoint` field on the `Cluster` object to be set before starting to provision the control plane.
```

Reviewer: This is outdated, please disregard

>   - As soon as the control plane endpoint is available, the controller must create a `ClusterEndpoint` CR to communicate the control plane endpoint to the Cluster API core controllers

Reviewer 2: ```suggestion
  - As soon as the Managed Kubernetes Service provided control plane endpoint is available, the controller must create a `ClusterEndpoint` CR to communicate this to the Cluster API core controllers
```

>   - As soon as the control plane endpoint is available, the controller must create a `ClusterEndpoint` CR to communicate the control plane endpoint to the Cluster API core controllers

Reviewer: Have we considered the possibility that the user will create this CR? What should the behavior be if the CR already exists?

Reviewer 6: > Have we considered the possibility that the user will create this CR?

This probably has to be a supported use case right? Given we position this CRD as a way to replace the controlPlaneEndpoint fields in InfraCluster CRDs and in InfraCluster users today can use this field to set a controlPlaneEndpoint

>   - The `ClusterEndpoint` CR must have an owner reference to the `ControlPlane` object from which is originated.

Reviewer 5: should this section have a recommendation for how machine controllers should position for infraCluster ommission?

Author: cc @fabriziopandini

Reviewer 4: I'm not sure I fully got the question, but I don't think so because managed Kubernetes use machine pools, which in turn have MachinePoolsMachines (which are "just placeholders" for machines created and managed by something else, not by CAPI controllers)

Reviewer 5: > but I don't think so because managed Kubernetes use machine pools

I think that's possibly the most common use case but we are not imposing that at the API. We are making a building block optional. Something like recommending provider machine controllers to be self API driven and approach infraCluster as a composable piece rather than a hard req imposition would be positive imo.

- Must implement a `FKRControlControlplane provider`, a `FKRControlControlplane` CRD, the corresponding `FKRControlControlplane` and related controllers
  - The `FKRControlControlplane` controller:
    - Must wait for `status.InfrastructureReady` field on the `Cluster` object to be set to true before starting to provision the `FKS` managed Kubernetes instance.
    - Must wait for `spec.ControlPlaneEndpoint` field on the `Cluster` object to be set before starting to provision the `FKS` managed Kubernetes instance.
    - Must set the `status.Ready` field on the `FKRControlControlplane` object when the provisioning is complete

Please note that this scenario is equivalent to what is implemented for a non managed Kubernetes `FooCluster`, backed by Cluster API managed `FooMachines`, with the only difference that in this case it possible to rely on `KCP` as `ControlControlplane provider`, and thus point 2 of the above list do not apply.

## Background work

This proposal builds on top of the awesome research work of the Managed Kubernetes working group, and it is a result of a huge work of a team of passionate Cluster API contributors.

> Below a summary of the main evidences / alternative considered during this this work.

Reviewer 2: Do we need this considering the original proposal has been updated along with this PR?

Author: I agree with you that the information starting here is not helping clarify things (and possibly needs more cleaning up). I've asked @fabriziopandini if he thinks we can just remove it.

Reviewer 4: Up to you folks, you did most of the background work, 
if you think that this doesn't help to clarify this proposal we can move it to a linked document (or let this implicit in the link to the updated original proposal)

Author: I'll remove for clarity. This was an attempt to restate the history that is already documented in other forms. In retrospect, asking someone to re-read this history again is too much.

This flow would require making changes to CAPI controllers so that there is an option to report the `ControlPlaneEndpoint` via the `<Infra>ControlPlane` as an alternative to coming from the `<Infra>Cluster`.

Using CAPG as an example:

```go
type GCPManagedControlPlaneSpec struct {
  // AddonsConfig defines the addons to enable with the GKE cluster.
  // +optional
  AddonsConfig *AddonsConfig `json:"addonsConfig,omitempty"`

  // Logging contains the logging configuration for the GKE cluster.
  // +optional
  Logging *ControlPlaneLoggingSpec `json:"logging,omitempty"`

  // EnableKubernetesAlpha will indicate the kubernetes alpha features are enabled
  // +optional
  EnableKubernetesAlpha bool

  // ControlPlaneEndpoint represents the endpoint used to communicate with the control plane.
  // +optional

>   ControlPlaneEndpoint clusterv1.APIEndpoint `json:"controlPlaneEndpoint"`

Reviewer 6: This seems to be outdated, now that we introduced the ClusterEndpoint CRD

Author: I actually think everything after the `## Background work` sub-heading can be removed. With the evolved proposal, all of this text is just too much IMO.

WDYT @fabriziopandini?

Reviewer 4: answered in https://github.com/kubernetes-sigs/cluster-api/pull/8500/files#r1378834165
(ok to drop if that's ok for you)

Author: removed

- Differentiates "self-managed clusters" from "managed clusters" at the foundational API layer:
  - Self-managed clusters would use the `Cluster` API resource as the top-level primitive object
  - Managed clusters would use the `ManagedCluster` API resource as the top-level primitive object
  - For example, to see all clusters under management at present, you can issue a `kubectl get clusters --all-namespaces` command (or the API equivalent); going forward, you would issue `kubectl get clusters,managedclusters --all-namespaces`
- There are no existing provider implementations. All existing provider implementations (e.g., CAPA, CAPZ, CAPOCI, CAPG) would need to be replaced or augmented in order to use a new spec.

## Recommendations

Because Managed Kubernetes was not yet in scope for Cluster API when it first appeared and gained rapid adoption, we are incentivized for paths forward that use the existing, mature, widely used API specification. The option to create a new `ManagedCluster` API type to best enforce provider consistency thusly has a high bar to clear in order to justify itself as the best option for the next phase of Managed Kubernetes in Cluster API.

> We conclude that enabling the the CAPI controllers to source authoritative `ControlPlaneEndpoint` data from the `<Infra>ControlPlane` resource is non-invasive to existing API contracts, and offers non-trivial flexibility for CAPI Managed Kubernetes providers at a small additional cost to CAPI maintenance going forward. Existing implementations that leverage a "proxy" `<Infra>Cluster` resource merely to satisfy CAPI contracts can be simplified by dropping the `<Infra>Cluster` resource altogether at little-to-no cost to their existing user communities. New Managed Kubernetes provider implementations will now have a little more flexibility to use a implementation that uses only a `<Infra>ControlPlane` resource, if that is appropriate, or for implementations that define both a `<Infra>Cluster` + `<Infra>ControlPlane` with the appropriate configuration distribution [following our recommendation][managedKubernetesRecommendation], those implementations can be non-trivially simplified with their `ControlPlaneEndpoint` data being observed and straightforwardly returned via `<Infra>ControlPlane`, the most common source of truth for a Managed Kubernetes service.

Reviewer 5: typo "enabling the the CAPI"

Author: this section has been removed as it describes a previous proposal recommendation

#### Core Cluster API changes

This proposal does not introduce any breaking changes for the existing "core" API. More specifically:

The existing Cluster API types are already able to omit the `<Infra>Cluster`:

- The `infrastructureRef` field on the Cluster object is already a pointer and thus it could be set to nil, and in fact we are already creating Clusters without `infrastructureRef` when we use a cluster class).
- The `infrastructure.Ref` field on the ClusterClass objects already a pointer and thus it could be set to nil, but in this case it is required to change the validation webhook to allow the user to not specify it; on top of that, when validating inline patches, we should reject patches targeting the infrastructure template objects if not specified.

In order to allow the `ControlPlane Provider` component to take ownership of the responsibility of creating the control plane endpoint we are going to introduce a new `ClusterEndpoint` CRD, below some example:

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterEndpoint
spec:
  host: "name-1234567890.region.elb.amazonaws.com"
  port: 1234
  type: ExternalControlPlaneEndpoint
```

```yaml

> apiVersion: cluster.x-k8s.io/v1beta1
> kind: ClusterEndpoint
> spec:
>   host: "10.40.85.102"
>   port: 1234
>   type: ExternalControlPlaneEndpoint

Reviewer 4: ```suggestion
apiVersion: cluster.x-k8s.io/v1beta1
kind: ClusterEndpoint
metadata:
   labels:
      cluster.x-k8s.io/cluster-name: my-cluster
spec:
  cluster: my-cluster
  host: "10.40.85.102"
  port: 1234
  type: ExternalControlPlaneEndpoint
```
Add label and field for cluster (we have them in all the CRs linked to a cluster)

Author: done

```

This is how the type specification would look:

```go
// ClusterEndpoint represents a reachable Kubernetes API endpoint serving a particular cluster function.

> type ClusterEndpoint struct {

Reviewer 4: If I'm not wrong we are missing the cluster field (added above) and the type field

Author: What are our final thoughts on proposing including a type field? There has been conversation for both options, including it now or doing it later...

Author: Updated with a type field, PTAL. Open to removing it if we want that to happen later in the evolution of this new CR.
